src/main/java/org/mule/extensions/java/api/exception/ArgumentMismatchModuleException.java

```diff
@@ -22,25 +30,88 @@
  */
 public class ArgumentMismatchModuleException extends JavaModuleException {
 
-  public ArgumentMismatchModuleException(String failure, Executable executable, Map<String, TypedValue<Object>> args) {
-    super(buildMessage(failure, executable, toHumanReadableArgs(args)), ARGUMENTS_MISMATCH);
+  public ArgumentMismatchModuleException(String failure, Executable executable,
+                                         Map<String, TypedValue<Object>> args,
+                                         ParametersTransformationResult transformationResult) {
+    super(buildMessage(failure, executable, args, transformationResult, empty()), ARGUMENTS_MISMATCH);
   }
 
-  public ArgumentMismatchModuleException(String failure, Executable executable, List<Object> args, Throwable cause) {
-    super(buildMessage(failure, executable, toHumanReadableArgs(args)) + ": " + cause.getMessage(),
+  public ArgumentMismatchModuleException(String failure, Executable executable,
+                                         Map<String, TypedValue<Object>> args,
+                                         ParametersTransformationResult transformationResult,
+                                         Throwable cause) {
+    super(buildMessage(failure, executable, args, transformationResult, getCauseMessage(cause)),
           ARGUMENTS_MISMATCH, cause);
   }
 
-  public ArgumentMismatchModuleException(String failure, Executable executable,
-                                         Map<String, TypedValue<Object>> args, Throwable cause) {
-    super(buildMessage(failure, executable, toHumanReadableArgs(args)) + ": " + cause.getMessage(),
-          ARGUMENTS_MISMATCH, cause);
+  private static String buildMessage(String failure, Executable executable,
+                                     Map<String, TypedValue<Object>> args,
+                                     ParametersTransformationResult transformationResult,
+                                     Optional<String> cause) {
+    StringBuilder sb = new StringBuilder()
+        .append(failure).append(". ");
+
+    Parameter[] parameters = executable.getParameters();
+    sb.append("Expected arguments are ")
+        .append(toHumanReadableArgs(parameters))
```

Comment:
what happens when parameters is a 0 size array?

Reply:
results in "[]", which can be better. I'll change it